Sources/BraintreeCoreSwift/BTAnalyticsService.swift

```diff
@@ -0,0 +1,175 @@
+    let sessionsQueue: DispatchQueue = DispatchQueue(label: "com.braintreepayments.BTAnalyticsService")
+
+    /// The HTTP client for communication with the analytics service endpoint. Exposed for testing.
+    var http: BTHTTP?
+
+    /// Defaults to 1, can be overridden
+    var flushThreshold: Int
+
+    /// Dictionary of analytics sessions, keyed by session ID. The analytics service requires that batched events
+    /// are sent from only one session. In practice, BTAPIClient.metadata.sessionID should never change, so this
+    /// is defensive.
+    var analyticsSessions: [String: BTAnalyticsSession] = [:]
+
+    private let apiClient: BTAPIClient
+
+    // MARK: - Initializer
+
+    init(apiClient: BTAPIClient, flushThreshold: Int = 1) {
+        self.apiClient = apiClient
+        self.flushThreshold = flushThreshold
+    }
+
+    // MARK: - Internal Methods
+
+    ///  Tracks an event.
+    ///
+    ///  Events are queued and sent in batches to the analytics service, based on the status of the app and
+    ///  the number of queued events. After exiting this method, there is no guarantee that the event has been sent.
+    /// - Parameter eventName: String representing the event name
+    func sendAnalyticsEvent(_ eventName: String) {
+        DispatchQueue.main.async {
+            self.enqueueEvent(eventName)
+            self.flushIfAtThreshold()
+        }
+    }
+
+    func sendAnalyticsEvent(_ eventName: String, completion: @escaping (Error?) -> Void = { _ in }) {
+        DispatchQueue.main.async {
+            self.enqueueEvent(eventName)
+            self.flush(completion)
+        }
+    }
+
+    func flush(_ completion: @escaping (Error?) -> Void = { _ in }) {
+        apiClient.fetchOrReturnRemoteConfiguration { configuration, error in
+            guard let configuration, error == nil else {
+                if let error {
+                    completion(error)
+                }
+                return
+            }
+
+            guard let analyticsURL = configuration.json?["analytics"]["url"].asURL() else {
+                completion(BTAnalyticsServiceError.missingAnalyticsURL)
+                return
+            }
+
+            if self.http == nil {
+                if let clientToken = self.apiClient.clientToken {
+                    self.http = BTHTTP(url: analyticsURL, authorizationFingerprint: clientToken.authorizationFingerprint)
+                } else if let tokenizationKey = self.apiClient.tokenizationKey {
+                    self.http = BTHTTP(url: analyticsURL, tokenizationKey: tokenizationKey)
+                } else {
+                    completion(BTAnalyticsServiceError.invalidAPIClient)
+                    return
+                }
+            }
+
+            // A special value passed in by unit tests to prevent BTHTTP from actually posting
+            if self.http?.baseURL.absoluteString == "test://do-not-send.url" {
```

Comment:
_Nitpick_ - instead of `if self.http == nil` on line 65 we could do a guard let, so that we can remove the optional chaining here.